Fix calendar view for CronTriggerTimeTable dags

bbovenzi opened the pull request:
We were checking the wrong instance when building the calendar data for CronTrigger dags and then getting stuck in an infinite loop.


Fixes: https://github.com/apache/airflow/issues/27645

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

bbovenzi committed:
check correct isinstance for crontriggertimetable dags

uranusjr commented on airflow/www/views.py:
```diff
@@ -2828,7 +2828,7 @@ def _convert_to_date(session, column):
             dates = collections.Counter()
 
-            if isinstance(dag.timetable, CronDataIntervalTimetable):
+            if isinstance(dag.timetable, CronTriggerTimetable):
```
Hmm, actually I think this was intended for CronDataIntervalTimetable, since the code was initially an optimisation to make rendering faster, and actually predates CronTriggerTimetable.

So I think we should do two things

1. _Add_ CronTriggerTimetable to the check (not replace)
2. Figure out why CronTriggerTimetable does not work with the `else` branch and fix that.

uranusjr replied:
This logic was added in #24262; CronTriggerTimetable was added in #25503 roughly a month later.

bbovenzi replied:
1. Included both cron timetables in the if statement

2. `next_dagrun_info` was always the same [here](https://github.com/apache/airflow/blob/main/airflow/www/views.py#L2844). So we never reached either of the `break` statements.

uranusjr replied:
Hmm, that does not sound right, the timetable should always be progressing forward. @mai-nakagawa maybe you could spot what’s wrong here?

(In the mean time I’m going to add a check to prevent infinite loop from happening)

uranusjr replied:
Ah I get it, because CronTriggerTimetable always uses the current time when catchup=False without considering `last_automated_data_interval`. It should; I’ll fix it in a different PR. (Edit: filed in #28532)

ashb replied:
@uranusjr Does your PR mean we don't need this PR at all?

uranusjr replied:
This would still serve as an optimisation.

bbovenzi committed:
check for both cron timetables

uranusjr committed:
Add guard to while loop to prevent infinite loops

The next DAG run schedule returned by a timetable should always be in
the "future" (later than the previous run), so we should simply check
that to prevent an incorrectly implemented timetable from breaking the
UI entirely.
